### PR TITLE
[Issue 715] - Add images to the building model

### LINKED ISSUE #715: Add images to the building model
A building can have multiple images associated with it. The model needs to allow adding images to a building with the following specifications:

1. file type: the files should have a file type restriction (they can only be PNG/JPEG);
2. compression: lossy compression with a target quality of **85% for JPEG and 80% for PNG**;
3. number of files: should allow a limit to the number of images uploaded; for now, set the limit to 3 images per building.

The specifications should allow for easy customization (e.g. change the number of allowed images, change the target quality).
Images should be uploadable through the admin interface and the API (the `/public_create` endpoint).

### PR DESCRIPTION
**Pull request by calindragomir**
### What does it fix?
Adds Images model and support for admin interface.

Closes #715 

### How has it been tested?

Add a few images locally, test out limits imposed.

#### Known issues
- left some comments in code for reviewer's decision
- no external image support
- no unit tests for this specific change

### CODE REVIEW

File: api/buildings/models.py [hunk 1]
```diff
@@ -248,3 +257,72 @@ class Meta:
...
+
+        # after modifications, save it to the output
+        im.save(
+            output,
+            format=str(self.image_type).upper(),
+            quality=settings.QUALITY_DEFINITIONS[str(self.image_type).lower()],
+        )
+        output.seek(0)
+
+        # change the imagefield value to be the newley modifed image value
```
Comment (tudoramariei) [inline api/buildings/models.py:318]: *newly modified

File: api/buildings/models.py [hunk 2]
```diff
@@ -248,3 +257,72 @@ class Meta:
...
+    ]
+
+    def set_images_limit(building):
+        if ImageFile.objects.filter(building_id=building).count() >= settings.ALLOWED_IMAGES_LIMIT:
+            raise ValidationError("Image limit for building is reached (%s)" % settings.ALLOWED_IMAGES_LIMIT)
+
+    def image_thumb(self):
+        return mark_safe(
+            '<a href={0}><img src="{0}" url width="50" height="50" /></a>'.format(
+                os.path.join(settings.MEDIA_URL, str(self.image))
```
Comment (danniel) [inline api/buildings/models.py:285]: Why not `self.image.url` ?

File: api/buildings/models.py [hunk 3]
```diff
@@ -248,3 +257,72 @@ class Meta:
...
+        im = PIL.Image.open(self.image)
+
+        output = BytesIO()
+
+        # TODO: do we want to resize it in any way ?
+        # Resize/modify the image
+        # im = im.resize((500, 500))
+
+        # after modifications, save it to the output
+        im.save(
```
Comment (danniel) [inline api/buildings/models.py:311]: The image processing should take place only if the image file was changed, not on each model save.

File: api/buildings/models.py [hunk 4]
```diff
@@ -248,3 +257,72 @@ class Meta:
...
+    ACCEPTED = 1
+    REJECTED = -1
+
+    IMAGE_STATUS_CHOICES = [
+        (PENDING, _("Pending")),
+        (ACCEPTED, _("Accepted")),
+        (REJECTED, _("Rejected")),
+    ]
+
+    def set_images_limit(building):
```
Comment (danniel) [inline api/buildings/models.py:278]: I think that a name like `check_image_limit` would be more clear about this function's purpose.

File: api/buildings/models.py [hunk 5]
```diff
@@ -248,3 +257,72 @@ class Meta:
...
+        if ImageFile.objects.filter(building_id=building).count() >= settings.ALLOWED_IMAGES_LIMIT:
```
Comment (danniel) [inline api/buildings/models.py:279]: `filter(building=building)`

File: api/buildings/admin.py [hunk 6]
```diff
@@ -27,6 +35,38 @@ def has_add_permission(self, request):
         return False
 
 
+@admin.register(models.ImageFile)
+class Images(admin.ModelAdmin):
```
[the remark this answers is not in the document]
Comment (calindragomir) [reply api/buildings/admin.py:39]: Yes, renamed.

File: api/buildings/admin.py [hunk 7]
```diff
@@ -27,6 +35,38 @@ def has_add_permission(self, request):
...
+    list_display = ("name", "image_type", "image_thumb")
```
[the remark this answers is not in the document]
Comment (calindragomir) [reply api/buildings/admin.py:40]: Updated, added `status` and removed `image_type`.

File: api/buildings/admin.py [hunk 8]
```diff
@@ -27,6 +35,38 @@ def has_add_permission(self, request):
...
+
+    def make_accepted(self, request, queryset):
+        self._perform_status_change(request, queryset, "1")
+
+    make_accepted.short_description = _("Mark selected images as accepted")
+
+    def make_rejected(self, request, queryset):
+        self._perform_status_change(request, queryset, "-1")
+
+    make_rejected.short_description = _("Mark selected images as rejected")
```
[the remark this answers is not in the document]
Comment (calindragomir) [reply api/buildings/admin.py:55]: I've made the necessary changes to get this to work.

File: api/buildings/models.py (hunk 4 above)
Comment (calindragomir) [reply api/buildings/models.py:278]: Good suggestion, I've updated the name.

File: api/buildings/models.py (hunk 5 above)
Comment (calindragomir) [reply api/buildings/models.py:279]: Updated.

File: api/buildings/models.py (hunk 2 above)
Comment (calindragomir) [reply api/buildings/models.py:285]: I was not aware of this attribute, this is now fixed.

File: api/buildings/models.py [hunk 9]
```diff
@@ -248,3 +257,72 @@ class Meta:
...
+            '<a href={0}><img src="{0}" url width="50" height="50" /></a>'.format(
+                os.path.join(settings.MEDIA_URL, str(self.image))
+            )
+        )
+
+    image_thumb.short_description = "Thumbnail"
+
+    building = models.ForeignKey(Building, validators=(set_images_limit,), on_delete=models.CASCADE)
+    name = models.CharField(_("name"), max_length=255)
+    image_type = models.CharField(_("type"), max_length=255, choices=settings.ACCEPTED_IMAGE_TYPES)
```
[the remark this answers is not in the document]
Comment (calindragomir) [reply api/buildings/models.py:293]: I've removed this from the DB, validation is done now exclusively using the filename. Three options are accepted: `jpg, jpeg and png` and the list is easy to expand only via the settings file.

File: api/buildings/models.py [hunk 10]
```diff
@@ -248,3 +257,72 @@ class Meta:
...
+
+    def save(self):
+        # Opening the uploaded image
+        im = PIL.Image.open(self.image)
+
+        output = BytesIO()
+
+        # TODO: do we want to resize it in any way ?
+        # Resize/modify the image
+        # im = im.resize((500, 500))
```
[the remark this answers is not in the document]
Comment (calindragomir) [reply api/buildings/models.py:308]: This is calculated dynamically now and I've added an unit test to validate the logic.

File: api/buildings/models.py (hunk 3 above)
Comment (calindragomir) [reply api/buildings/models.py:311]: Good catch, I've updated the code to make this possible.

File: api/buildings/models.py (hunk 1 above)
Comment (calindragomir) [reply api/buildings/models.py:318]: My bad, this is fixed.

**Review by calindragomir — COMMENTED** [review]
Thanks @tudoramariei and @danniel for your reviews, I'm pushing a new commit with the fixes implemented.

File: api/seismic_site/settings.py [hunk 11]
```diff
@@ -151,6 +151,11 @@ class Base(Configuration):
     HERE_MAPS_API_KEY = os.getenv("HERE_MAPS_API_KEY")
     HERE_MAPS = {"api_key": HERE_MAPS_API_KEY}
 
+    ACCEPTED_IMAGE_TYPES = {"JPEG": ["jpg", "jpeg"], "PNG": ["png"]}
```
[the remark this answers is not in the document]
Comment (calindragomir) [reply api/seismic_site/settings.py:154]: done

**Review by tudoramariei — APPROVED** [review]

**Pull request merged**